### ISSUE TITLE
SWA compared to Exponential Moving Average (EMA)

### ISSUE THREAD

**Issue by glenn-jocher** [timgaripov/swa#16]
Hello, I recently discovered your PyTorch SWA implementation. Have you compared it to the EMA implementation that TF uses?
https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage

EfficientDet uses EMA with a decay of 0.9998 during training, and I recently implemented it for https://github.com/ultralytics/yolov3 with mixed [results](https://github.com/rwightman/pytorch-image-models/issues/102#issuecomment-601424476).

Any comments would be appreciated!

**Comment by glenn-jocher** [comment]
There is also a PyTorch EMA implementation at https://github.com/rwightman/pytorch-image-models. See https://github.com/rwightman/pytorch-image-models/issues/106

**Comment by izmailovpavel** [comment]
Hey @glenn-jocher thanks for the question. Yes, we did compare to EMA, and SWA does better in our experiments. A big part of SWA is the learning rate schedule, we use a high learning rate late in training to ensure that SGD explores multiple good models. Check out our blogpost https://pytorch.org/blog/stochastic-weight-averaging-in-pytorch/#is-this-just-averaged-sgd for more detail.

**Comment by mrgloom** [comment]
Is my understanding right that:

EMA is doing `shadow_variable = decay * shadow_variable + (1 - decay) * variable` for each train step https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage

And SWA is doing something like `avr_model_weights = (model1_weights + model2_weights + ... + modelN_weights) / N`, where `model{i}_weights` is taken from lowest lr point in cyclic lr shedule at each cycle?

**Comment by izmailovpavel** [comment]
Hey @mrgloom. Yeah, that's correct, SWA is a regular average of the SGD iterates, and we are using a modified lr schedule. Note that you don't have to use the cyclical schedule, and you can often get good results with just using a constant learning rate in SWA.
